pkg/skaffold/build/cache/retrieve.go

```diff
@@ -52,86 +56,70 @@ func (c *Cache) RetrieveCachedArtifacts(ctx context.Context, out io.Writer, arti
 	start := time.Now()
 	color.Default.Fprintln(out, "Checking cache...")
 
+	detailsErrs := make([]chan detailsErr, len(artifacts))
+
+	for i := range artifacts {
+		detailsErrs[i] = make(chan detailsErr, 1)
+
+		i := i
+		go func() {
+			details, err := c.retrieveCachedArtifactDetails(ctx, artifacts[i])
+			detailsErrs[i] <- detailsErr{details: details, err: err}
+		}()
+	}
+
 	var (
 		needToBuild []*latest.Artifact
 		built       []build.Artifact
-
-		wg   sync.WaitGroup
-		lock sync.Mutex
 	)
 
-	wg.Add(len(artifacts))
+	for i, artifact := range artifacts {
+		color.Default.Fprintf(out, " - %s: ", artifact.ImageName)
 
-	for _, a := range artifacts {
-		a := a
-		go func() {
-			defer wg.Done()
-
-			artifact, err := c.resolveCachedArtifact(ctx, out, a)
+		select {
+		case <-ctx.Done():
+			return nil, nil, context.Canceled
 
-			lock.Lock()
-			defer lock.Unlock()
+		case d := <-detailsErrs[i]:
```

Comment:
Is there a chance we could ever get here before detailsErrs[i] has been saved?

Reply:
No, the array is initialised in the previous loop, in the main go routine.